public/TCP/IMC.js

```diff
@@ -0,0 +1,514 @@
+const encode = {
+  estimatedState: encodeEstimatedState,
+  entityState: encodeEntityState,
+  desiredControl: encodeDesiredControl,
+  lowLevelControlManeuver: {
+    desiredHeading: encodeLowLevelControlManeuverDesiredHeading,
+    desiredZ: encodeLowLevelControlManeuverDesiredZ,
+  },
+  goTo: encodeGoTo,
+  netFollow: encodeNetFollow,
+  customNetFollow: encodeCustomNetFollowState,
+};
+
+function decode(buf) {
+  let id = buf.readUInt16BE(0);
+  return decodeImc(buf, idToMessageMetadata[id]);
+}
+
+const datatypes = {
+  uint_8t: {
+    name: 'uint_8t',
+    length: 1,
+  },
+  uint_16t: {
+    name: 'uint_16t',
+    length: 2,
+  },
+  uint_32t: {
+    name: 'uint_16t',
+    length: 4,
+  },
+  fp32_t: {
+    name: 'fp32_t',
+    length: 4,
+  },
+  fp64_t: {
+    name: 'fp64_t',
+    length: 8,
+  },
+  bitfield: {
+    name: 'bitfield',
+    length: 1,
+  },
+  recursive: {
+    name: 'recursive',
+  },
+};
+
+function encodeIMC(imcMessage, imcMessageMetadata) {
+  // Check if assumed length is corret
+  if (
+    imcMessageMetadata.length !==
+    imcMessageMetadata.id.datatype.length +
+      lenImcMessage(imcMessageMetadata.message)
+  ) {
+    console.log(
+      `WARNING: Length not equal to calculated length. Assumed length: ${
+        imcMessageMetadata.length
+      } calculated length: ${imcMessageMetadata.id.datatype.length +
+        lenImcMessage(imcMessageMetadata.message)}`,
+    );
+  }
+
+  let buf = Buffer.alloc(imcMessageMetadata.length);
+
+  // Encode ID
+  buf.writeUInt16BE(imcMessageMetadata.id.value);
+  let offset = imcMessageMetadata.id.datatype.length;
+
+  imcMessageMetadata.message.map(imcEntity => {
+    let imcValue = Object.prototype.hasOwnProperty.call(imcEntity, 'value')
+      ? imcEntity.value
+      : imcMessage[imcEntity['name']];
+    switch (imcEntity.datatype) {
+      case datatypes.uint_8t:
+        buf.writeUIntBE(imcValue, offset, imcEntity.datatype.length);
+        break;
+      case datatypes.uint_16t:
+        buf.writeUInt16BE(imcValue, offset);
+        break;
+      case datatypes.uint_32t:
+        buf.writeUInt32BE(imcValue, offset);
+        break;
+      case datatypes.fp32_t:
+        buf.writeFloatBE(imcValue, offset);
+        break;
+      case datatypes.fp64_t:
+        buf.writeDoubleBE(imcValue, offset);
+        break;
+      case datatypes.bitfield:
+        buf.writeUIntBE(
+          bitfieldToUIntBE(imcValue, imcEntity.fields),
+          offset,
+          datatypes.bitfield.length,
+        );
+        break;
+      default:
+        break;
+    }
+    offset += imcEntity.datatype.length;
+  });
+  return buf;
+}
+
+function decodeImc(buf, imcMessageMetadata) {
+  const result = {};
+  let offset = 2; // Do not need to decode id
```

Comment:
dont understand the comment

Reply:
I guess the two first bytes are the ID, so the decoding starts at byte 2 (offset=2) instead. But agree that the comment could be more explanatory 👌🏼